ls1intum/Artemis
# Add Dynamic Tests for the Artemis Domain Classes

### Checklist

### Motivation and Context
Currently, some domain class methods are not tested or the tests only print to the console.

We try to add proper tests that adapt to whatever domain classes we have and check some basic invariants and behavioral properties. The main focus are the methods of `java.lang.Object`.

We could also add tests to see if our classes are in fact `Serializable`.

### Description
- Added class path scanning utilities
- Added test class for dynamic test of the domain classes

#### Todo List
- [x] Add exceptions to some rules
- [x] Fix the bugs found by the tests
- [x] Remove `System.out.println` in all the other tests
- [ ] See if https://github.com/gradle/gradle/issues/5975 is an impediment?
       We have many tests with the same name, but different `DynamicContainers`. In the console, the path is not displayed correctly.

### Steps for Testing
1. Review the code, check that you can understand most of what is does.
2. Run the tests on your machine, especially the `ObjectMethodTest` in your IDE.

### Test Coverage


- +12 something according to Codacy
- +300 instructions according to JaCoCo
- **Now vs. Before**
  ![image](https://user-images.githubusercontent.com/11130248/95077826-69298080-0714-11eb-8c1d-c8f2a44fd246.png)

### Screenshots
![image](https://user-images.githubusercontent.com/11130248/95062495-55beeb00-06fd-11eb-9378-5449379f50b8.png)
